### PR TITLE
[Feat] Binary focal loss

### PR DESCRIPTION
**Pull request by xen0f0n**
New PR In reference to #699 #506 #493.

### CODE REVIEW

**Review by edgarriba — APPROVED** [review]
LGTM

**Review by ducha-aiki — APPROVED** [review]
lgtm

File: kornia/losses/focal.py [hunk 1]
```diff
@@ -122,3 +122,109 @@ def forward(  # type: ignore
...
+    args:
+        inp (torch.Tensor) : input data tensor
+        target (torch.Tensor): the target tensor
+        alpha (float): Weighting factor for the rare class :math:`\alpha \in [0, 1]`.  Default: 0.25.
+        gamma (float): Focusing parameter :math:`\gamma >= 0`.  Default: 2.0.
+        reduction (str, optional): Specifies the reduction to apply to the.  Default: 'none'.
+        eps (float): for numerically stability when dividing. Default: 1e-8.
+
+    returns:
+        torch.tensor: loss
```
Comment (shijianjian) [inline kornia/losses/focal.py:145]: Can you add a doctest to demonstrate that if the outputting values are desired?

Comment (xen0f0n) [reply kornia/losses/focal.py:145]: @shijianjian do you mean to post here the output of the doctest to the example in BinaryFocalLossWithLogits?

Examples:
        >>> N = 5  # num_classes
        >>> kwargs = {"alpha": 0.25, "gamma": 2.0, "reduction": 'mean'}
        >>> loss = BinaryFocalLossWithLogits(**kwargs)
        >>> input = torch.randn(1, N, 3, 5, requires_grad=True)
        >>> target = torch.empty(1, 3, 5, dtype=torch.long).random_(N)
        >>> output = loss(input, target)
        >>> output.backward()
    """

Comment (shijianjian) [reply kornia/losses/focal.py:145]: I mean to have something like to verify that it could output the exact loss value as expected:

```python
>>> logits = [0.1, 0.2, 0.3]
>>> label = [1, 2, 3]
>>> binary_focal_loss_with_logits(logits, label)
2.892
```

Comment (xen0f0n) [reply kornia/losses/focal.py:145]: @shijianjian have a look...

**Review by shijianjian — APPROVED** [review]
